src/core/thread/network_data_service.cpp

```diff
@@ -165,7 +172,8 @@ Error Manager::GetNextDnsSrpAnycastInfo(Iterator &aIterator, DnsSrpAnycast::Info
 
     do
     {
-        tlv = Get<Leader>().FindNextMatchingService(tlv, kThreadEnterpriseNumber, &serviceData, sizeof(serviceData));
+        tlv = Get<Leader>().FindNextMatchingService(tlv, kThreadEnterpriseNumber, &serviceData, sizeof(serviceData),
```

Comment:
Should we also relax the matching here as well?
Currently service data for `DNS-SD Anycast Dataset` is of length 2, but may also subject to have additional data in the future likewise.

Reply:
@librasungirl  Agree; that looks good now (using 'kServiceDataPrefixMatch') and that is per the specification (length may grow in the future).

Reply:
@librasungirl the existing code already did a partial match pf service data. `FindNextMatchingService()` was added to do partial service data check  from this PR and for purpose of using it for DNS/SRP service info:
- #6473

> This commit adds a new method in `NetworkData` class to find and
> iterate over all Service TLVs that start with a given Service Data.
> Unlike the existing `FindService()` method which searches for a
> Service TLV with an exact match with the given Service Data,  this
> method performs a relaxed check allowing a matching Service TLV to
> contain additional bytes after given data bytes to search for.
> 
> This commit also updates `test_network_data` unit test to add new
> test to cover the behavior of newly added method.
> 
> ----
> 
> This will be useful for parsing Service TLVs for SRP/DNS server info as
> discussed/described in https://github.com/openthread/openthread/issues/6447.  
>

Reply:
Agree that there is partial match already and Thread device could find the service of `new` format (with additional field).  
My proposal of using a constant of `2` is out of backward compatibility concern, future specification may add new field, but won't remove existing field. If someday a new field is introduced, and the thread end device of new version joins an existing Thread network with `old` service, we may still want it be able to recognize the `old` service and perform SRP.  However with `sizeof(serviceData)` approach, new device could only recognize `new` service.

Thoughts?

Reply:
@librasungirl Agree, an implementation should treat any size >=2 as valid in order to correctly recognize current-version service data and also future-version service data.   That means using sizeof() is risky for future implementations in case the default (new) service data size grows, because then that newer implementation won't recognize the 'old' size=2 data anymore which it should do.   It may be useful to check if this issue also occurs in other places.

Reply:
@librasungirl I think the existing code already does what you are proposing (i.e. it accepts `service data len >= size=2`) and should work fine if additional fields are added in future to service data (i.e., device running the current code would accept them as valid and ignore the additional data). 

Maybe I am missing something here. Let me describe the code:

```c++
    uint8_t           serviceData = DnsSrpAnycast::kServiceNumber;
    const ServiceTlv *tlv         = aIterator.mServiceTlv;

    do
    {
        tlv = Get<Leader>().FindNextMatchingService(tlv, kThreadEnterpriseNumber, &serviceData, sizeof(serviceData));
        VerifyOrExit(tlv != nullptr, error = kErrorNotFound);

    } while (tlv->GetServiceDataLength() < sizeof(DnsSrpAnycast::ServiceData));
```

- The `serviceData` we are using in `FindNextMatchingService()` is just a single byte, so `sizeof(serviceData)` is 1.
  - In the call to `FindNextMatchingService()` we are basically accepting any service where the first byte matches the `DnsSrpAnycast::kServiceNumber (0x5c)`.
  - We then later check if there are any other fields after the service number.
- The check in `while()` loop `tlv->GetServiceDataLength() < sizeof(DnsSrpAnycast::ServiceData)` ensures that the service data is at least the minimum length we expect (or `2`) - we skip over any matching service that does not have the minimum expected length (we currently know)
   - Of course, if in future we add more fields to service data,  the code needs to change to handle the new fields and handle the backward compatibility, e.g., future/new code adding one more byte to service data may accept both size 2 or 3 or more and in case of an entry we size of 2 we need to define how to set the missing info (init with some default value, etc,). This is up to new code/spec (when/if there is a change to service data format). 
   - But any device running the current code (i.e., old code) would continue to work correctly.
- I think we even cover the extra bytes in service data in the unit test `TestNetworkDataDsnSrpServices()` (i.e. we have an entry in Network Data example in this test with some additional bytes in service data after the 2 and we ensure that it is accepted and parsed correctly).
- Do you see any concern with this? Thanks.

Reply:
Thanks for the details. I overlooked `uint8_t           serviceData = DnsSrpAnycast::kServiceNumber;`  in previous diff review somehow. 
Agree that no issue on `FindNextMatchingService()`.  For `while()` loop, may consider using a constant 2 in comparison as it is known minimal length other than a variable `sizeof(class)` for best compatibility now. But as you pointed, when there is new fields, corresponding process should be updated as well, sgtm to leave it as is.

Reply:
Just noting for code below (line 171) it uses the expression sizeof(DnsSrpAnycast::ServiceData)  , which is then only valid as long as the service data length is not increased by the specifications. When the specification changes and so the size of the DnsSrpAnycast::ServiceData becomes longer, the sizeof() construct can no longer be used. That property seems to make it less useful to use a sizeof() construct.  I would assume sizeof() is handy for code that automatically keeps up with increases or decreases in the size of the given object, which is probably not the intended behaviour right here.

Reply:
Thanks @librasungirl , Thanks @EskoDijk.

My view on this is that if/when the definition of service data changes (new fields are added), it is the responsibility of new code to handle this. If the new code continues to use `sizeof()` then it is a bug in new code. 

We can define `kMinLength` today and use it in the code to try to protect for future changes but it does not add much protection since the new code needs to anyways handle the additional fields as optional (check against different lengths as parsing the data). 

The main reason I would say we, in general, prefer and use `sizeof()` over defining a constant value directly is that complier is better at getting the size over people (compliers hopefully makes less mistake 😄).